clang/runtime/dpct-rt/include/dpct/dpl_extras/dpcpp_extensions.h

```diff
@@ -566,6 +566,89 @@ class radix_sort {
   uint8_t *_local_memory;
 };
 
+/// Load linear segment items into block format across threads
+/// Helper for Block Load
+enum load_algorithm {
+
+  BLOCK_LOAD_DIRECT,
+  BLOCK_LOAD_STRIPED,
+  // To-do: BLOCK_LOAD_WARP_TRANSPOSE
+
+};
+
+// loads a linear segment of workgroup items into a subgroup striped
+// arrangement. Created as free function until exchange mechanism is
+// implemented.
+// To-do: inline this function with BLOCK_LOAD_WARP_TRANSPOSE mechanism
+template <size_t GROUP_WORK_ITEMS, size_t ITEMS_PER_WORK_ITEM, typename InputT,
+          typename InputIteratorT, typename Item>
+__dpct_inline__ void
+load_subgroup_striped(const Item &item, size_t linear_tid,
+                      InputIteratorT block_itr,
+                      InputT (&items)[ITEMS_PER_WORK_ITEM]) {
+
+  // This implementation does not take in account range loading across
+  // workgroup items To-do: Decide whether range loading is required for group
+  // loading
+
+  size_t subgroup_offset = item.get_sub_group().get_local_range()[0];
+#pragma unroll
+  for (uint32_t idx = 0; idx < ITEMS_PER_WORK_ITEM; idx++) {
+    new (&items[idx]) InputT(
+        block_itr[subgroup_offset + linear_tid + (idx * GROUP_WORK_ITEMS)]);
+  }
+}
+
+template <size_t GROUP_WORK_ITEMS, size_t ITEMS_PER_WORK_ITEM,
+          load_algorithm ALGORITHM, typename InputT, typename InputIteratorT,
+          typename Item>
+class workgroup_load {
+public:
+  workgroup_load(uint8_t *local_memory) : _local_memory(local_memory) {}
+
+  __dpct_inline__ void load(const Item &item, size_t linear_tid,
+                            InputIteratorT block_itr,
+                            InputT (&items)[ITEMS_PER_WORK_ITEM]) {
+
+    if constexpr (ALGORITHM == BLOCK_LOAD_DIRECT) {
+      load_blocked(linear_tid, block_itr, (&items)[ITEMS_PER_WORK_ITEM]);
+    }
+    if constexpr (ALGORITHM == BLOCK_LOAD_STRIPED) {
+      load_striped(linear_tid, block_itr, (&items)[ITEMS_PER_WORK_ITEM]);
+    }
+  }
+  // loads a linear segment of workgroup items into a blocked arrangement.
+  __dpct_inline__ void load_blocked(size_t linear_tid, InputIteratorT block_itr,
+                                    InputT (&items)[ITEMS_PER_WORK_ITEM]) {
+
+    // This implementation does not take in account range loading across
+    // workgroup items To-do: Decide whether range loading is required for group
+    // loading
+
+    uint32_t workgroup_offset = linear_tid * ITEMS_PER_WORK_ITEM;
+#pragma unroll
+    for (uint32_t idx = 0; idx < ITEMS_PER_WORK_ITEM; idx++) {
+      items[idx] = block_itr[workgroup_offset + idx];
+    }
+  }
+  // loads a linear segment of workgroup items into a striped arrangement.
+  __dpct_inline__ void load_striped(size_t linear_tid, InputIteratorT block_itr,
```

Comment:
Do we want to expose this publicly? If we do, it may be worth pulling it out as a a free function, so it can be invoked directly from here as well. It could still be called similarly within `load`.

If not, we should make it private. The same goes for `load_blocked`.

Reply:
Free function makes sense, can be used both ways.

Reply:
I agree that free functions make the most sense.